uinput.go

```diff
@@ -0,0 +1,177 @@
+	if err != nil {
+		return nil, err
+	}
+
+	newDev := &InputDevice{
+		file: deviceFile,
+	}
+
+	for ev, codes := range capabilities {
+		if err := ioctlUISETEVBIT(newDev.file.Fd(), uintptr(ev)); err != nil {
+			DestroyDevice(newDev)
+			return nil, fmt.Errorf("failed to set ev bit: %d", ev)
+		}
+
+		if err := setEventCodes(newDev, ev, codes); err != nil {
+			DestroyDevice(newDev)
+			return nil, fmt.Errorf("failed to set ev code")
+		}
+	}
+
+	_, err = createUsbDevice(newDev.file, UinputUserDevice{
+		Name: toUinputName([]byte(name)),
+		ID: InputID{
+			BusType: 0x03,
+			Vendor:  0x4712,
+			Product: 0x0816,
+			Version: 1,
+		},
+	})
+	if err != nil {
+		DestroyDevice(newDev)
+		return nil, err
+	}
+
+	return newDev, nil
+}
+
+// CloneDevice from an existing one
+// all capabilites will be coppied over to the new virtual device
+// If setup fails the device will be removed from the system,
+// once setup it can be removed by calling dev.Close
+func CloneDevice(dev *InputDevice) (*InputDevice, error) {
+	deviceFile, err := os.OpenFile("/dev/uinput", syscall.O_WRONLY|syscall.O_NONBLOCK, 0660)
+	if err != nil {
+		return nil, err
+	}
+
+	newDev := &InputDevice{
+		file:          deviceFile,
+		driverVersion: dev.driverVersion,
+	}
+
+	for _, ev := range dev.CapableTypes() {
+		if err := ioctlUISETEVBIT(newDev.file.Fd(), uintptr(ev)); err != nil {
+			DestroyDevice(newDev)
+			return nil, fmt.Errorf("failed to set ev bit: %d - %s", ev, err)
+		}
+
+		eventCodes := dev.CapableEvents(ev)
+		if err := setEventCodes(newDev, ev, eventCodes); err != nil {
+			DestroyDevice(newDev)
+			return nil, fmt.Errorf("failed to set ev code %s", err)
+		}
+	}
+
+	name, err := dev.Name()
+	if err != nil {
+		DestroyDevice(newDev)
+		return nil, errors.New("failed to get original device name")
+	}
+
+	id, err := dev.InputID()
+	if err != nil {
+		DestroyDevice(newDev)
+		return nil, errors.New("failed to get original device id")
+	}
+
+	_, err = createUsbDevice(newDev.file, UinputUserDevice{
+		Name: toUinputName([]byte(name + "(clone)")),
+		ID:   id,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return newDev, nil
+}
+
+func setEventCodes(dev *InputDevice, ev EvType, codes []EvCode) error {
+	for _, code := range codes {
+		var err error
+
+		switch ev {
+		case EV_ABS:
+			err = ioctlUISETABSBIT(dev.file.Fd(), uintptr(code))
+		case EV_FF:
+			err = ioctlUISETFFBIT(dev.file.Fd(), uintptr(code))
+		case EV_KEY:
+			err = ioctlUISETKEYBIT(dev.file.Fd(), uintptr(code))
+		case EV_LED:
+			err = ioctlUISETLEDBIT(dev.file.Fd(), uintptr(code))
+		case EV_MSC:
+			err = ioctlUISETMSCBIT(dev.file.Fd(), uintptr(code))
+		case EV_REL:
+			err = ioctlUISETRELBIT(dev.file.Fd(), uintptr(code))
+		case EV_SND:
+			err = ioctlUISETSNDBIT(dev.file.Fd(), uintptr(code))
+		case EV_SW:
+			err = ioctlUISETSWBIT(dev.file.Fd(), uintptr(code))
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
You can spare this extra `return`

Reply:
This is a simple enough change to make but it would change the logic of the function. It would have the potential to either allow 10s or even 100s of write calls to be made to a device that has already failed to set up and is doomed to be destroyed, or event worse if another set bit call succeeds afterwards it would mask the earlier error, causing unexpected behavior/missing capabilities once the device is setup.

Admittedly this would only happen in very edge case scenarios but its worth consideration.

Reply:
Yeah, sorry. I missed that. Leave it as is, it's fine!